src/cluster/websocket/WebSocket.cc

```diff
@@ -0,0 +1,480 @@
+    auto backend = zeek::cluster::manager->InstantiateBackend(cluster_backend_val, std::move(event_serializer), nullptr,
+                                                              std::move(event_handling_strategy));
+
+    WS_DEBUG("New WebSocket client %s (%s:%d) - using id %s backend=%p", id.c_str(), wsc->getRemoteIp().c_str(),
+             wsc->getRemotePort(), ws_id.c_str(), backend.get());
+
+    // XXX: We call InitPostScript to populate member vars required for connectivity.
+    backend->InitPostScript();
+    backend->Init(std::move(ws_id));
+
+    clients[id] = WebSocketClientEntry{id, wsc, std::move(backend)};
+}
+
+void WebSocketEventDispatcher::Process(WebSocketClose& close) {
+    const auto& id = close.id;
+    const auto& it = clients.find(id);
+
+    if ( it == clients.end() ) {
+        reporter->Error("Close from non-existing WebSocket client with id %s!", id.c_str());
+        return;
+    }
+
+    auto wsc = it->second.wsc;
+    auto& backend = it->second.backend;
+
+    WS_DEBUG("Close from client %s (%s:%d) backend=%p", wsc->getId().c_str(), wsc->getRemoteIp().c_str(),
+             wsc->getRemotePort(), backend.get());
+
+    auto rec = zeek::cluster::detail::bif::make_endpoint_info(backend->NodeId(), wsc->getRemoteIp(),
+                                                              wsc->getRemotePort(), TRANSPORT_TCP);
+    zeek::event_mgr.Enqueue(Cluster::websocket_client_lost, std::move(rec));
+
+    backend->Terminate();
+
+    clients.erase(it);
+}
+
+// SubscribeFinished is produced internally.
+void WebSocketEventDispatcher::Process(WebSocketSubscribeFinished& fin) {
+    const auto& it = clients.find(fin.id);
+    if ( it == clients.end() ) {
+        reporter->Error("Subscribe finished from non-existing WebSocket client with id %s!", fin.id.c_str());
+        return;
+    }
+
+    auto& entry = it->second;
+    auto& wsc = entry.wsc;
+
+    entry.wsc->SetSubscriptionActive(fin.topic_prefix);
+
+    if ( ! entry.wsc->AllSubscriptionsActive() ) {
+        // More subscriptions to come.
+        return;
+    }
+
+    auto rec = zeek::cluster::detail::bif::make_endpoint_info(backend->NodeId(), wsc->getRemoteIp(),
+                                                              wsc->getRemotePort(), TRANSPORT_TCP);
+    auto subscriptions_vec = zeek::cluster::detail::bif::make_string_vec(wsc->GetSubscriptions());
+    zeek::event_mgr.Enqueue(Cluster::websocket_client_added, std::move(rec), std::move(subscriptions_vec));
+
+    entry.wsc->SendAck(entry.backend->NodeId(), zeek::zeek_version());
+
+    WS_DEBUG("Sent Ack to %s %s\n", fin.id.c_str(), entry.backend->NodeId().c_str());
+
+    // Process any queued messages now.
+    for ( auto& msg : entry.queue ) {
+        assert(entry.msg_count > 1);
+        Process(msg);
+    }
+}
+
+void WebSocketEventDispatcher::HandleSubscriptions(WebSocketClientEntry& entry, std::string buf) {
+    rapidjson::Document doc;
+    doc.Parse(buf.c_str());
```

Comment:
Nit: No need to pass a string copy.

```suggestion
void WebSocketEventDispatcher::HandleSubscriptions(WebSocketClientEntry& entry, std::string_view buf) {
    rapidjson::Document doc;
    doc.Parse(buf.data(), buf.size());
```